### DISCUSSION TITLE
vs. rust htslib

### DISCUSSION THREAD

**Discussion by Kentalot** [zaeleus/noodles#207] [General]
Hi,

I was looking at using rust for bioinformatics processing, but was wondering what the main difference between noodles and rust-htslib is from a practical user perspective?  Is noodles supposed to be faster or simply pure rust or anything?  htslib is used widely and well tested so seems like [rust-htslib](https://github.com/rust-bio/rust-htslib) can do almost everything noodles can do, no?  Is noodles significantly faster or anything?  I am aware that rust-htslib doesn't have bed functionality, though they do have rust-bio that does do beds?  Any comments would be appreciated, especially by the authors.  Thanks!

**Comment by zaeleus** [comment]
The primary motivation of noodles is to attempt to provide compliant implementations of well-known large scale genomics specifications, particularly [hts-specs]. This is quite different from existing implementations, which tend to read and write particular dialects of the formats, either for historical reasons or ignorance of the specifications. As such, it hinders interoperability, reducing confidence when sharing data or collaborating among users of different tools.

When strictly dealing with I/O, yes, htslib and noodles are rather similar. Both can read and write records in the supported alignment (SAM, BAM, and CRAM) and variant (VCF and BCF) formats, and both allow random access using an index.

The decoders/parsers, however, have different ideologies. htslib is very lenient; whereas, noodles is far more strict. For example, take the following alignment record ([`207.bam`]):

```
@HD	VN:1.6
🍜	4	*	-7	255	4B	*	-7	0	ACGT	}~��	æ:i:0	æ:i:1
```

I'd argue that this is not a valid record. The read name is non-ASCII, the positions and quality scores are out of range, the CIGAR operation is invalid, and the auxiliary data tags are both invalid and duplicated. Yet, htslib/samtools will read (and rewrite) this without error.

```
$ samtools version | head -1
samtools 1.18
$ samtools view --no-PG --with-header 207.bam
@HD     VN:1.6
🍜      4       *       -7      255     4B      *       -7      0       ACGT    }~      æ:i:0   æ:i:1
$ echo $?
0
```

The typical record reader in noodles (i.e., `bam::Reader::read_record`) will fail reading this, stopping at the first invalid field.

```
$ cargo run --example bam_view 207.bam
Error: invalid position

Caused by:
    0: invalid input
    1: out of range integral type conversion attempted
```

But perhaps this fine; perhaps users want nonsensical and/or bad data. htslib made a choice to prefer performance over conformance. (noodles also provides an alternative lazy/raw record reader, which allows for per field decoding but does not allow rewriting them.) This is likely the major practical difference users will encounter in noodles when compared to htslib, i.e., specification stringency.

Some extra notes:

  * noodles' public API is less terse. They're not identical in usage, but see [`add_header.c`] vs [`sam_reheader.rs`], [`read_bam.c`] vs [`bam_view.rs`], etc. rust-htslib does have the added benefit of providing a nice facade over htslib's C API though.
  * Nearly all the crates in noodles are pure Rust implementations. noodles-cram relies on `libbz2` and `liblzma`, and noodles-bgzf can be optionally be built with `libdeflate`.
  * Positions in noodles are 1-based ([start, end]), and htslib is 0-based ([start, end)).
  * noodles should be competitive with htslib in performance, but remember that their decoders/parsers are not performing the same work.

[hts-specs]: https://samtools.github.io/hts-specs/
[`207.bam`]: https://github.com/zaeleus/noodles/files/12895702/207.bam.gz
[`add_header.c`]: https://github.com/samtools/htslib/blob/99415e2a2ce26bdbf4e910954330ea769de2c3f0/samples/add_header.c
[`sam_reheader.rs`]: https://github.com/zaeleus/noodles/blob/9dd09ff35d466a7bf315efa0f7c5f811bca13db8/noodles-sam/examples/sam_reheader.rs
[`read_bam.c`]: https://github.com/samtools/htslib/blob/99415e2a2ce26bdbf4e910954330ea769de2c3f0/samples/read_bam.c
[`bam_view.rs`]: https://github.com/zaeleus/noodles/blob/9dd09ff35d466a7bf315efa0f7c5f811bca13db8/noodles-bam/examples/bam_view.rs

**Comment by Kentalot** [comment]
Great, thanks for the detailed reply.  This makes sense and I think people do care a lot about valid data.  I will evaluate our needs and decide which one works best for our project.  Thanks so much!

**Comment by rLannes** [comment]
I have a question reading this, why choose 1-based position? This crate is going to be used by bioinformatician or computational biologist, not wet lab scientist. I was looking for an alternative to rust hts-lib. But 1-based position is a big friction/pain point for me.

**Reply by zaeleus** [reply]
Position coordinates in genomic regions are defined to be 1-based, e.g., chr1:1.

When working with a mixture of coordinate systems, 1-based positions guarantee arithmetic stability when converting (n - 1 >= 0); 0-based does not (n + 1 overflows MAX when n = MAX). This is an important consideration when choosing which system to use for normalization.

In noodles, owned sequences have access to the underlying data, so you can always continue to use 0-based indexing for those, e.g.,


```rust
let sequence = Sequence::from(b"ACGT");

// 1-based [start, end]
let (start, end) = const { (Position::new(2).unwrap(), Position::new(3).unwrap()) };
assert_eq!(sequence[start], b'C');
assert_eq!(&sequence[start..=end], b"CG");

// 0-based [start, end)
let raw_sequence = sequence.as_ref();
assert_eq!(raw_sequence[1], b'C');
assert_eq!(&raw_sequence[1..3], b"CG");
```
